### PR TITLE
Add more areas to strict null

### PR DESCRIPTION
**Pull request by anthonydresser**

### CODE REVIEW

File: src/sql/platform/backup/common/backupService.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export const IBackupService = createDecorator<IBackupService>(SERVICE_ID);
 export interface IBackupService {
 	_serviceBrand: undefined;
 
-	getBackupConfigInfo(connectionUri: string): Thenable<azdata.BackupConfigInfo>;
+	getBackupConfigInfo(connectionUri: string): Promise<azdata.BackupConfigInfo | void>;
```
Comment (anthonydresser) [inline src/sql/platform/backup/common/backupService.ts]: In general we should switch to using Promise everywhere since mixing thenable and promise can complicate things from a code flow point of view, and theres really now point in abstracting this out. The only exception would be in our api layer were we don't want to enforce a PromiseLike implementation on extensions (Plus there are multiple different, contradicting, implementation of Promise) .

File: src/sql/platform/backup/common/backupServiceImp.ts [hunk 2]
```diff
@@ -27,15 +27,15 @@ export class BackupService implements IBackupService {
 	/**
 	 * Get database metadata needed to populate backup UI
 	 */
-	public getBackupConfigInfo(connectionUri: string): Thenable<azdata.BackupConfigInfo> {
+	public getBackupConfigInfo(connectionUri: string): Promise<azdata.BackupConfigInfo | void> {
```
Comment (Charles-Gagnon) [inline src/sql/platform/backup/common/backupServiceImp.ts]: >void [](start = 86, length = 4)

void here seems odd - why did you change it from undefined? It seems like it's more correct to have this return some value all the time since the caller presumably wanted the actual value when they called it

File: src/sql/platform/fileBrowser/common/fileBrowserService.ts [hunk 3]
```diff
@@ -136,18 +136,19 @@ export class FileBrowserService implements IFileBrowserService {
 		this._onPathValidate.fire(fileBrowserValidatedParams);
 	}
 
-	public closeFileBrowser(ownerUri: string): Thenable<azdata.FileBrowserCloseResponse> {
+	public closeFileBrowser(ownerUri: string): Thenable<azdata.FileBrowserCloseResponse | void> {
```
Comment (Charles-Gagnon) [inline src/sql/platform/fileBrowser/common/fileBrowserService.ts]: >Thenable [](start = 44, length = 8)

Promise?

File: src/sql/platform/metadata/common/metadataService.ts [hunk 4]
```diff
@@ -14,13 +14,13 @@ export const IMetadataService = createDecorator<IMetadataService>(SERVICE_ID);
 export interface IMetadataService {
 	_serviceBrand: undefined;
 
-	getMetadata(connectionUri: string): Thenable<azdata.ProviderMetadata>;
+	getMetadata(connectionUri: string): Thenable<azdata.ProviderMetadata | void>;
```
Comment (Charles-Gagnon) [inline src/sql/platform/metadata/common/metadataService.ts]: >Thenable [](start = 37, length = 8)

Promise for the ones in this file as well?

File: src/sql/platform/tasks/browser/tasksRegistry.ts [hunk 5]
```diff
@@ -74,17 +74,19 @@ export abstract class Task {
...
 		this.id = opts.id;
 		this.title = opts.title;
-		this.iconPath = {
-			dark: opts.iconPath ? URI.parse(opts.iconPath.dark) : undefined,
-			light: opts.iconPath ? URI.parse(opts.iconPath.light) : undefined,
-		};
+		if (opts.iconPath.dark) {
+			this.iconPath = {
+				dark: URI.parse(opts.iconPath.dark),
+				light: opts.iconPath ? URI.parse(opts.iconPath.light) : undefined,
```
Comment (Charles-Gagnon) [inline src/sql/platform/tasks/browser/tasksRegistry.ts]: >opts.iconPath [](start = 11, length = 13)

You already use opts.iconPath above so this check isn't doing anything

File: src/sql/platform/metadata/common/metadataService.ts (hunk 4 above)
Comment (Charles-Gagnon) [reply src/sql/platform/metadata/common/metadataService.ts]: Others in the other service files too if you wanted to fix them

---
In reply to: [325389256](https://github.com/microsoft/azuredatastudio/pull/7243#discussion_r325389256) [](ancestors = 325389256)

File: src/sql/platform/backup/common/backupServiceImp.ts (hunk 2 above)
Comment (anthonydresser) [reply src/sql/platform/backup/common/backupServiceImp.ts]: Technically void is the absent of a type, where as undefined and null are actually types. so something like Promise.resolve(undefined) would be undefined, and Promise.resolve() would be void. Its more pedantics than anything. https://www.typescriptlang.org/docs/handbook/basic-types.html#void

Comment (anthonydresser) [reply src/sql/platform/backup/common/backupServiceImp.ts]: Now in reality javascript treats lack of a value as undefined. Personally when it comes to function I like seeing void if the function isn't going to return anything. It makes more sense when you use async function.

````typescript
async function foo(): Promise<string | void> {
     if(something) {
           return 'some string';
     }
}
````

looks better than me to
````typescript
async function foo(): Promise<string | undefined> {
     if(something) {
           return 'some string';
     }
     return undefined;
}

but they would do the same thing functionally.

Comment (anthonydresser) [reply src/sql/platform/backup/common/backupServiceImp.ts]: Although thinking about it when writing a normal function we generally do

````typescript
function foo(): string | undefined {
     if(something) {
           return 'some string';
     }
     return undefined;
}
````

so idk.

Comment (anthonydresser) [reply src/sql/platform/backup/common/backupServiceImp.ts]: Maybe the rule should be if you don't ever return anything use `void` if you somethings return anything use `something|undefined` so in the promise land, if your promise never has a value use `Promise<void>`. If your promise sometimes returns something use `Promise<something| undefined>`

Comment (anthonydresser) [reply src/sql/platform/backup/common/backupServiceImp.ts]: But like i said javascript functionality will consider `void` like typescript types as undefined so we are really pulling straws at this point.

Comment (Charles-Gagnon) [reply src/sql/platform/backup/common/backupServiceImp.ts]: Yeah - so I guess I wasn't sure the behavior of the return value from a void function. Coming from a managed world like C# though string | undefined feels much more natural to me. I also feel like that's the style we tend to use throughout the codebase currently (undefined if no/invalid value).

In the end though I'm fine adopting this if that's agreed to be the way going forward - I just want us to be consistent with usage if we can.

Comment (Charles-Gagnon) [reply src/sql/platform/backup/common/backupServiceImp.ts]: Yeah that would be my general preference. `void` only if the function doesn't ever return anything, otherwise `type | undefined` if it does at least some of the time.

Comment (aaomidi) [reply src/sql/platform/backup/common/backupServiceImp.ts]: We should also be looking at using nulls where it makes sense too. We're using null interchangeably with undefined even though they're not entirely the same thing.

But I guess since the typescript team doesn't use null themselves they probably have a good reason.

Comment (Charles-Gagnon) [reply src/sql/platform/backup/common/backupServiceImp.ts]: What advantages do you see to using null? We've been moving away from null for a while now and I don't really see any benefit in using that over undefined in nearly all cases.

Comment (anthonydresser) [reply src/sql/platform/backup/common/backupServiceImp.ts]: Technically they (pedantically) 2 different things, null is a empty value (like null pointer) where as undefined is a lack of any value. In reality we aren't dealing in areas where that difference matters. The one place we have to care is when interacting with the DOM which something like `element.style.color = undefined` will throw an error in strict null checks because the value is defined in typescript as `string | null`.
